File: Python/41.缺失的第一个正数.py
#
# @lc app=leetcode.cn id=41 lang=python3
#
# [41] 缺失的第一个正数
#
from typing import List
# @lc code=start
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        res = 0
        n = len(nums)

        for i in range(n):
            if nums[i] <= 0:
                nums[i] = n + 1

        for i in range(n):
            abs_num = abs(nums[i])

            if abs_num > 0 and abs_num <= n and nums[abs_num - 1] > 0:
                nums[abs_num - 1] *= -1

        while res < n and nums[res] < 0:
            res += 1

        return res + 1
        
# @lc code=end

File: Java/41.缺失的第一个正数.java
/*
 * @lc app=leetcode.cn id=41 lang=java
 *
 * [41] 缺失的第一个正数
 */

// @lc code=start
class Solution {
    public int firstMissingPositive(int[] nums) {
        int res = 0, n = nums.length;

        for (int i = 0; i < n; i++) {
            if(nums[i] <= 0)
                nums[i] = n + 1;
        }

        for (int i = 0; i < n; i++) {
            int absNum = Math.abs(nums[i]);

            if (absNum > 0 && absNum <= n && nums[absNum - 1] > 0)
                nums[absNum - 1] *= -1;
        }
        
        while (res < n && nums[res] < 0)
            res++;

        return res + 1;
    }
}
// @lc code=end

